ceno_zkvm/src/instructions/global.rs

```diff
@@ -0,0 +1,817 @@
+                // to ensure read => y in [0, p/2) and write => y in [p/2, p)
+                let negate = match (self.is_write, is_y_in_2nd_half) {
+                    (true, false) => true, // write, y in [0, p/2)
+                    (false, true) => true, // read, y in [p/2, p)
+                    _ => false,
+                };
+
+                let point = if negate { -p } else { p };
+
+                return GlobalPoint { nonce, point };
+            } else {
+                // try again with different nonce
+                nonce += 1;
+                input[6] = E::BaseField::from_canonical_u32(nonce);
+            }
+        }
+    }
+}
+/// opcode circuit + mem init/final table + local finalize circuit + global chip
+/// global chip is used to ensure the **local** reads and writes produced by
+/// opcode circuits / memory init / memory finalize table / local finalize circuit
+/// can balance out.
+///
+/// 1. For a local memory read record whose previous write is not in the same shard,
+///    the global chip will read it from the **global set** and insert a local write record.
+/// 2. For a local memory write record which will **not** be read in the future,
+///    the local finalize circuit will consume it by inserting a local read record.
+/// 3. For a local memory write record which will be read in the future,
+///    the global chip will insert a local read record and write it to the **global set**.
+pub struct GlobalConfig<E: ExtensionField, P> {
+    addr: WitIn,
+    is_ram_register: WitIn,
+    value: UInt<E>,
+    shard: WitIn,
+    global_clk: WitIn,
+    local_clk: WitIn,
+    nonce: WitIn,
+    // if it's a write to global set, then insert a local read record
+    // s.t. local offline memory checking can cancel out
+    // this serves as propagating local write to global.
+    is_global_write: WitIn,
+    x: Vec<WitIn>,
+    y: Vec<WitIn>,
+    slope: Vec<WitIn>,
+    perm_config: Poseidon2Config<E, 16, 7, 1, 4, 13>,
+    perm: P,
+}
+
+impl<E: ExtensionField, P> GlobalConfig<E, P> {
+    // TODO: make `WIDTH`, `HALF_FULL_ROUNDS`, `PARTIAL_ROUNDS` generic parameters
+    pub fn configure(
+        cb: &mut CircuitBuilder<E>,
+        rc: RoundConstants<E::BaseField, 16, 4, 13>,
+        perm: P,
+    ) -> Result<Self, CircuitBuilderError> {
+        let x: Vec<WitIn> = (0..SEPTIC_EXTENSION_DEGREE)
+            .map(|i| cb.create_witin(|| format!("x{}", i)))
+            .collect();
+        let y: Vec<WitIn> = (0..SEPTIC_EXTENSION_DEGREE)
+            .map(|i| cb.create_witin(|| format!("y{}", i)))
+            .collect();
+        let slope: Vec<WitIn> = (0..SEPTIC_EXTENSION_DEGREE)
+            .map(|i| cb.create_witin(|| format!("slope{}", i)))
+            .collect();
+        let addr = cb.create_witin(|| "addr");
+        let is_ram_register = cb.create_witin(|| "is_ram_register");
+        let value = UInt::new(|| "value", cb)?;
+        let shard = cb.create_witin(|| "shard");
```

Comment:
follow comments here https://github.com/scroll-tech/ceno/pull/1061#discussion_r2472029220
we can remove `shard` and just keep `global_clk`